packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/BuddyBridge.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/*
+ * Copyright 2021, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity ^0.6.11;
+
+import "arb-bridge-eth/contracts/bridge/interfaces/IInbox.sol";
+import "arb-bridge-eth/contracts/bridge/interfaces/IOutbox.sol";
+import "arb-bridge-eth/contracts/bridge/interfaces/IBridge.sol";
+
+import "../arbitrum/BuddyDeployer.sol";
+
+// contracts that want to have buddies should inherit from this
+abstract contract BuddyContract {
+    bool public connected;
+    L2Deployer public l2Deployer;
+    IInbox public inbox;
+    bytes32 codeHash;
+
+    constructor(address _inbox, address _l2Deployer) public {
+        connected = false;
+        inbox = IInbox(_inbox);
+        l2Deployer = L2Deployer(_l2Deployer);
+    }
+
+    function initiateBuddyDeploy(
+        uint256 maxGas,
+        uint256 gasPriceBid,
+        bytes calldata deployCode
+    ) external payable {
+        require(!connected, "already connected");
+        // deployCode == type(ArbSymmetricTokenBridge).creationCode
+        bytes memory data = abi.encodeWithSelector(L2Deployer.executeBuddyDeploy.selector, deployCode);
+
+        if(msg.value > 0) {
+            // gas paid in L1
+            inbox.sendL1FundedContractTransaction(maxGas, gasPriceBid, address(l2Deployer), data);
+        } else {
+            // gas paid in L2
+            inbox.sendContractTransaction(maxGas, gasPriceBid, address(l2Deployer), 0, data);
+        }
+        codeHash = keccak256(deployCode);
+    }
+
+    function finalizeBuddyDeploy(
+        bool success
+    ) external {
+        // get sender from outbox
+        IOutbox outbox = IOutbox(inbox.bridge().activeOutbox());
+
+        address expectedL2Address = calculateL2Address(
+            address(l2Deployer),
+            address(this),
+            codeHash
+        );
+        require(outbox.l2ToL1Sender() == expectedL2Address, "Wrong L2 address triggering outbox");
```

Comment:
This will be the `L2Deployer` and not the deployed L2 contract I think

Reply:
Pushed a commit with the following logic:
If deployed, the call comes from L2 deployed contract constructor
If failed, the call comes from the L2 Deployer